node/src/types/block.rs

```diff
@@ -1111,6 +1113,217 @@ impl Item for BlockHeaderWithMetadata {
+        if &highest_hash != latest_known.parent_hash() {
+            return Err(BlockHeadersBatchValidationError::HighestBlockHashMismatch {
+                expected: *latest_known.parent_hash(),
+                got: highest_hash,
+            });
+        }
+
+        if !Self::is_continuous_and_descending(self.inner(), verifiable_chunked_hash_activation) {
+            return Err(BlockHeadersBatchValidationError::BatchNotContinuous);
+        }
+
+        self.0
+            .last()
+            .cloned()
+            .ok_or(BlockHeadersBatchValidationError::BatchEmpty)
+    }
+
+    /// Tries to create an instance of `BlockHeadersBatch` from a `Vec<BlockHeader>`.
+    ///
+    /// Returns `Some(Self)` if data passes validation, otherwise `None`.
+    pub fn from_vec(
+        batch: Vec<BlockHeader>,
+        requested_id: &BlockHeadersBatchId,
+        verifiable_chunked_hash_activation: EraId,
+    ) -> Option<Self> {
+        match batch.first() {
+            Some(highest) => {
+                if highest.height() != requested_id.highest {
+                    error!(expected_highest=?requested_id.highest, got_highest=?highest, "unexpected highest block header");
+                    return None;
+                }
+            }
+            None => {
+                error!("response cannot be an empty batch");
+                return None;
+            }
+        }
+
+        match batch.last() {
+            Some(lowest) => {
+                if lowest.height() != requested_id.lowest {
+                    error!(expected_lowest=?requested_id.lowest, got_lowest=?lowest, "unexpected lowest block header");
+                    return None;
+                }
+            }
+            None => {
+                error!("input cannot be empty");
+                return None;
+            }
+        }
+
+        if !Self::is_continuous_and_descending(&batch, verifiable_chunked_hash_activation) {
+            error!("batch is not continuous");
+            return None;
+        }
+
+        Some(Self(batch))
+    }
+
+    /// Returns inner value.
+    pub fn into_inner(self) -> Vec<BlockHeader> {
+        self.0
+    }
+
+    /// Returns a reference to an inner vector of block headers.
+    pub fn inner(&self) -> &Vec<BlockHeader> {
+        &self.0
+    }
+
+    /// Tests whether the block header batch is continuous and in descending order.
+    pub fn is_continuous_and_descending(
+        batch: &[BlockHeader],
+        verifiable_chunked_hash_activation: EraId,
+    ) -> bool {
+        batch
+            .windows(2)
+            .filter_map(|window| match &window {
+                &[l, r] => Some((l, r)),
+                _ => None,
+            })
+            .all(|(l, r)| {
+                l.height() == r.height() + 1
+                    && l.parent_hash() == &r.hash(verifiable_chunked_hash_activation)
+            })
+    }
+
+    #[cfg(test)]
+    // Test-only constructor allowing creation otherwise invalid data.
```

Comment:
```suggestion
    // Test-only constructor allowing creation of otherwise invalid data.
```

Reply:
Done in [`60ac7e8` (#3020)](https://github.com/casper-network/casper-node/pull/3020/commits/60ac7e8598c652d5e8d043c149df8319bdd03027)